libyara/grammar.y

```diff
@@ -1722,14 +1612,63 @@ expression
     ;
 
 
+iterator
+    : identifier
+      {
+        // Initially we assume that the identifier is from a non-iterable type,
+        // this will change later if it's iterable.
+        int result = ERROR_WRONG_TYPE;
+
+        if ($1.type == EXPRESSION_TYPE_OBJECT)
+        {
+          switch($1.value.object->type)
+          {
+            case OBJECT_TYPE_ARRAY:
+              $$.type = EXPRESSION_TYPE_OBJECT;
+              $$.value.object = object_as_array($1.value.object)->prototype_item;
+              result = yr_parser_emit(yyscanner, OP_ITER_START_ARRAY, NULL);
+              break;
+          }
+        }
+
+        if (result == ERROR_WRONG_TYPE)
+        {
+          yr_compiler_set_error_extra_info_fmt(
+              compiler,
+              "identifier \"%s\" is not iterable",
+              $1.identifier);
+        }
+
+        fail_if_error(result);
+      }
+    | integer_set
+      {
+        $$.type = EXPRESSION_TYPE_INTEGER;
+        $$.value.integer = UNDEFINED;
+      }
+    ;
+
+
 integer_set
-    : '(' integer_enumeration ')'  { $$ = INTEGER_SET_ENUMERATION; }
-    | range                        { $$ = INTEGER_SET_RANGE; }
+    : '(' integer_enumeration ')'
```

Comment:
I assume you are intentionally not using string enumerations here? In theory they could be treated as an iterator, right?

Reply:
That's correct. I'm leaving that for another commit in order to keep this one as small as possible.

Reply:
That makes sense. Thanks for clarifying!